statsmodels/tools/data_interface.py

```diff
@@ -0,0 +1,136 @@
+import pandas as pd
+import numpy as np
+from functools import partial
+from patsy import dmatrix
+
+NUMPY_TYPES = [np.ndarray, np.float64]
+
+class DataInterface(object):
+
+    def __init__(self, permitted_types, internal_type, data=None, external_type=None, model=None, use_formula=False):
+
+        self.permitted_types = permitted_types
+        self.internal_type = internal_type
+        self.columns = getattr(data, 'columns', None)
+        self.name = getattr(data, 'name', None)
+        self.index = getattr(data, 'index', None)
```

Comment:
This is a bit dangerous/fragile: You use the index from the data in the `__init__`, but the actual data in the methods could have a different `index`.